sdk/mobile-center-updates/src/main/java/com/microsoft/azure/mobile/updates/Updates.java

```diff
@@ -0,0 +1,971 @@
+            DownloadManager.Request request = new DownloadManager.Request(downloadUrl);
+            long downloadRequestId = downloadManager.enqueue(request);
+            storeDownloadRequestId(downloadManager, this, downloadRequestId);
+            return null;
+        }
+    }
+
+    /**
+     * Inspect a completed download, this uses APIs that would trigger strict mode violation if used in U.I. thread.
+     */
+    @VisibleForTesting
+    class ProcessDownloadCompletionTask extends AsyncTask<Void, Void, Void> {
+
+        /**
+         * Context.
+         */
+        private final Context mContext;
+
+        /**
+         * Download identifier to inspect.
+         */
+        private final long mDownloadId;
+
+        /**
+         * Init.
+         *
+         * @param context    context.
+         * @param downloadId download identifier.
+         */
+        ProcessDownloadCompletionTask(Context context, long downloadId) {
+            mContext = context;
+            mDownloadId = downloadId;
+        }
+
+        @Override
+        protected Void doInBackground(Void... params) {
+
+            /*
+             * Completion might be triggered in background before MobileCenter.start
+             * if application was killed after starting download.
+             *
+             * We still want to generate the notification: if we can find the data in preferences
+             * that means they were not deleted, and thus that the sdk was not disabled.
+             */
+            MobileCenterLog.debug(LOG_TAG, "Process download completion id=" + mDownloadId);
+            if (Updates.this.mContext == null) {
+                MobileCenterLog.debug(LOG_TAG, "Called before onStart, init storage");
+                StorageHelper.initialize(mContext);
+            }
+
+            /* Check intent data is what we expected. */
+            long expectedDownloadId = StorageHelper.PreferencesStorage.getLong(PREFERENCE_KEY_DOWNLOAD_ID);
+            if (expectedDownloadId > 0 && expectedDownloadId != mDownloadId) {
+                MobileCenterLog.warn(LOG_TAG, "Ignoring completion for a download we didn't expect, id=" + mDownloadId);
+                return null;
+            }
+
+            /* Check if download successful. */
+            DownloadManager downloadManager = (DownloadManager) mContext.getSystemService(DOWNLOAD_SERVICE);
+            Uri uriForDownloadedFile = downloadManager.getUriForDownloadedFile(mDownloadId);
+            if (uriForDownloadedFile != null) {
+
+                /* Build install intent. */
+                MobileCenterLog.debug(LOG_TAG, "Download was successful for id=" + mDownloadId + " uri=" + uriForDownloadedFile);
+                Intent intent = getInstallIntent(uriForDownloadedFile);
+                boolean installerFound = false;
+                if (intent.resolveActivity(mContext.getPackageManager()) == null) {
```

Comment:
You never know and startActivity throws ActivityNotFoundException if you don't guard.